parser/src/ast/declarations.rs

```diff
@@ -0,0 +1,523 @@
+                AccessType::Slice(range) if range.end > len => {
+                    Err(InvalidAccessError::IndexOutOfBounds)
+                }
+                AccessType::Slice(range) => Ok(Self::Vector(range.end - range.start)),
+                AccessType::Index(idx) if idx >= len => Err(InvalidAccessError::IndexOutOfBounds),
+                AccessType::Index(_) => Ok(Self::Integer),
+                AccessType::Matrix(_, _) => Err(InvalidAccessError::IndexIntoScalar),
+                _ => unreachable!(),
+            },
+            Self::Matrix(rows, cols) => match access_type {
+                AccessType::Slice(range) if range.end > rows => {
+                    Err(InvalidAccessError::IndexOutOfBounds)
+                }
+                AccessType::Slice(range) => Ok(Self::Matrix(range.end - range.start, cols)),
+                AccessType::Index(idx) if idx >= rows => Err(InvalidAccessError::IndexOutOfBounds),
+                AccessType::Index(_) => Ok(Self::Vector(cols)),
+                AccessType::Matrix(row, col) if row >= rows || col >= cols => {
+                    Err(InvalidAccessError::IndexOutOfBounds)
+                }
+                AccessType::Matrix(_, _) => Ok(Self::Integer),
+                _ => unreachable!(),
+            },
+        }
+    }
+}
+impl fmt::Display for Type {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::Integer => f.write_str("integer"),
+            Self::Vector(n) => write!(f, "vector of length {}", n),
+            Self::Matrix(rows, cols) => write!(f, "matrix of {} rows and {} columns", rows, cols),
+        }
+    }
+}
+
+/// An import declaration
+///
+/// There can be multiple of these in a given module
+#[derive(Debug, Clone)]
+pub enum Import {
+    /// Imports all items from `module`
+    All { module: ModuleId },
+    /// Imports `items` from `module`
+    Partial {
+        module: ModuleId,
+        items: HashSet<Identifier>,
+    },
+}
+impl Import {
+    pub fn module(&self) -> ModuleId {
+        match self {
+            Self::All { module } | Self::Partial { module, .. } => *module,
+        }
+    }
+}
+impl Eq for Import {}
+impl PartialEq for Import {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::All { module: l }, Self::All { module: r }) => l == r,
+            (
+                Self::Partial {
+                    module: l,
+                    items: ls,
+                },
+                Self::Partial {
+                    module: r,
+                    items: rs,
+                },
+            ) if l == r => ls.difference(rs).next().is_none(),
+            _ => false,
+        }
+    }
+}
+
+/// Represents an item exported from a module
+///
+/// Currently, only constants and functions are exported.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Export<'a> {
+    Constant(&'a crate::ast::Constant),
+    Evaluator(&'a EvaluatorFunction),
+}
```

Comment:
We should be able to export pure functions as well (but maybe those just aren't here yet)

Reply:
Yep, that's just not added here because there is no `Function` type in the AST yet, I assumed we'd add that immediately after these modularity changes. The changes in this PR do anticipate functions though, and are prepped for when such a type is added.